### ISSUE TITLE
delete-by-query plugin triggers CircuitBreakingException / SearchContextMissingException on large queries

### ISSUE THREAD

**Issue by j16r** [elastic/elasticsearch#18411]
**Elasticsearch version**:

5.0.0~alpha2

**JVM version**:

java version "1.8.0_91"
Java(TM) SE Runtime Environment (build 1.8.0_91-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.91-b14, mixed mode)

**OS version**:

Linux ip-10-10-155-12 3.13.0-74-generic #118-Ubuntu SMP Thu Dec 17 22:52:10 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**:

**Steps to reproduce**:
 1. Create a large number of documents
 2. Attempt to delete the documents using something like a match_all query and the delete-by-query plugin
 3. Boom!

**Provide logs (if relevant)**:

You may see any combination of these:

```
[2016-05-15 00:38:29,675][ERROR][action.deletebyquery     ] [client] scroll request [...] failed, scrolling document(s) is stopped
Failed to execute phase [query], all shards failed; shardFailures {RemoteTransportException[[worker][10.10.155.231:9300][indices:data/read/search[phase/query/scroll]]]; nested: SearchContextMissingException[No search context found for id [3894]]; }{RemoteTransportException[[worker][10.10.155.184:9300][indices:data/read/search[phase/query/scroll]]]; nested: SearchContextMissingException[No search context found for id [3894]]; }{RemoteTransportException[[worker][10.10.155.83:9300][indices:data/read/search[phase/query/scroll]]]; nested: SearchContextMissingException[No search context found for id [3897]]; }{RemoteTransportException[[worker][10.10.155.181:9300][indices:data/read/search[phase/query/scroll]]]; nested: SearchContextMissingException[No search context found for id [3894]]; }{RemoteTransportException[[worker][10.10.155.192:9300][indices:data/read/search[phase/query/scroll]]]; nested: SearchContextMissingException[No search context found for id [3897]]; }
        at org.elasticsearch.action.search.SearchScrollQueryThenFetchAsyncAction.onQueryPhaseFailure(SearchScrollQueryThenFetchAsyncAction.java:155)
        at org.elasticsearch.action.search.SearchScrollQueryThenFetchAsyncAction$1.onFailure(SearchScrollQueryThenFetchAsyncAction.java:142)
        at org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:51)
        at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:795)
        at org.elasticsearch.transport.netty.MessageChannelHandler.handleException(MessageChannelHandler.java:204)
        at org.elasticsearch.transport.netty.MessageChannelHandler.handlerResponseError(MessageChannelHandler.java:194)
        at org.elasticsearch.transport.netty.MessageChannelHandler.messageReceived(MessageChannelHandler.java:141)
        at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
        at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
        at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
        at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
        at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462)
        at org.jboss.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443)
        at org.jboss.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:303)
        at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
        at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
        at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
        at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
        at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
        at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
        at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
        at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:337)
        at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
        at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
        at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
        at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
[ERROR][action.deletebyquery     ] [client] scroll request [...] failed, scrolling document(s) is stopped
```

or:

```
2016-05-15 23:38:55,644WARNrest.suppressed /_bulk Params: {}
CircuitBreakingException[parent Data too large, data for <http_request> would be larger than limit of 2994274304/2.7gb]
at org.elasticsearch.indices.breaker.HierarchyCircuitBreakerService.checkParentLimit(HierarchyCircuitBreakerService.java:211)
at org.elasticsearch.common.breaker.ChildMemoryCircuitBreaker.addEstimateBytesAndMaybeBreak(ChildMemoryCircuitBreaker.java:128)
...
```

Yup, 2.7 gigs!

Since I had quite a few of these queries going in tandem, you can guess what happened next. Several of the worker nodes pegged at 200% CPU and started to use up all swap and ram (44gigs for the elasticsearch process in most cases). Searches would take 15 minutes. 

Looking at the task list I see a big backlog of scroll requests and delete-by-query tasks, none of them cancellable. Not sure what I can do here except kill the index and start again.

Looking quickly at https://github.com/elastic/elasticsearch/blob/master/plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/TransportDeleteByQueryAction.java I'm guessing that the plugin just builds a bulk delete request without considering how large it might get.

I think some first steps might be:
  1. Make this plugin cancellable, not sure what this involves but there seems plenty of opportunity in between pages of the scroll request to check a cancellation flag.
  2. At least error out before attempting to build queries of such a large size, the search loop could probably check the `BulkRequest#estimatedSizeInBytes` against the cluster's max request size limits.
  3. Chunk bulk requests into smaller components and deliver them piecemeal

**Comment by dakrone** [comment]
The circuit breaker issue is a duplicate of https://github.com/elastic/elasticsearch/issues/18144 that will be fixed by https://github.com/elastic/elasticsearch/pull/18204 (in alpha3)

**Comment by nik9000** [comment]
It is hard to tell without a curl recreation, but do you happen to be setting the `size` on the request? It looks like that sets the batch size.

**Comment by nik9000** [comment]
> It is hard to tell without a curl recreation, but do you happen to be setting the size on the request? It looks like that sets the batch size.

For what it is worth that won't be true after #18329 is merged and released. That'll make delete-by-query work just like update-by-query so `size` will actually be a size limit. And it'll default to unlimited.

**Comment by j16r** [comment]
@nik9000 just reading: https://github.com/elastic/elasticsearch/blob/230697c20220937ae5ffed4de29af73b5124d56d/modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkByScrollAction.java it doesn't look like the plugins will attempt to avoid creating gigantic bulk requests, only truncate them if you specify a size. Is my reading correct? 

Thankfully that PR seems to add cancellation capability!

**Comment by nik9000** [comment]
@j16r right. If you specify a massive batch_size parameter then it'll make a massive bulk request. I could certainly add validation for the batch size and forbid sizes above 50,000 or something just like we do with the result window on a regular query.

Do you know if your issue with delete-by-query was caused by setting a huge batch size?

**Comment by j16r** [comment]
I have not explicitly set a batch size, so it seems that it's defaulting to something quite large (unlimited?). It does not appear to be specified with the API I'm using: https://github.com/olivere/elastic/blob/v3.0.37/delete_by_query.go

**Comment by nik9000** [comment]
I had a look and I can't actually set a batch size over the REST api. Which means this is probably all the fault of #18329. It looks like delete-by-query is always using a batch size of 10, the default for searches. I expect porting it to reindex's infra will probably speed it up quite a bit then.

I think we're ok here - probably it is already fixed by #18204 which will come out with alpha 3. There isn't really a work around until then though.

Maybe we should add a preflight check to the _reindex family of requests to make sure the batch size is not crazy large?

**Comment by clintongormley** [comment]
I tested this out with size 1000 and size 5000, with sort:_doc and with sort:_uid.  There was little difference in performance (while 1000 gives a huge performance boost over 10).  

> Maybe we should add a preflight check to the _reindex family of requests to make sure the batch size is not crazy large?

Agreed.  Perhaps an upper limit of 10000?

**Comment by nik9000** [comment]
> Maybe we should add a preflight check to the _reindex family of requests to make sure the batch size is not crazy large?

We have an in-flight check for this in master (#19367), failing with batch sizes of greater than 10,000. I think we're ok here.
